### DISCUSSION TITLE
Failed to update prerender cache for /index

### DISCUSSION THREAD

**Discussion by mo-alaa** [vercel/next.js#51164] [Help]
### Summary

Hi! I have a problem in the logs. I am implementing ISR with `getStaticProps`.
For the exact pages the have ISR, I found this error: 

> Failed to update prerender cache for /index [Error: EACCES: permission denied, open '/app/.next/server/pages/index.html'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/app/.next/server/pages/index.html'
}

I use Docker, with a similar multi-stage deployment like in [this example](https://github.com/vercel/next.js/tree/canary/examples/with-docker)

I tried several approaches, like: 

- Upgrading nextJs to the latest version.
- Copying the whole next directory in the last build stage, runner (instead of the static directory only)
`COPY --from=builder --chown=nextjs:nodejs /app/.next ./.next`
- I created the pages directory and gave permissions write (during the runner stage with the root user)
`RUN mkdir -p /app/.next/server/pages`
`RUN chmod 777 /app/.next/server/pages`

- I even created the html page and gave it users permissions to write and execute, till my last build stage (runner) looked like this: 
![image](https://github.com/vercel/next.js/assets/41386480/12e46de1-aa9b-499d-807e-59e0c4b9bd5c)

Do you have any idea what is the cause of the error or how it can be resolved ?
Thanks!

**Comment by devjiwonchoi** [comment]
Try `User node` since the `node` user is provided in the Node.js Alpine base image.

## Reference
- #16995

**Reply by mo-alaa** [reply]
@devjiwonchoi Hmm It didn't work. However, the only thing that removed the error was giving the server directory recursive read/write/exec permissions for others like so: 
`RUN mkdir -p /app/.next/server`
`RUN chmod -R 777 /app/.next/server`

The weird thing is that 770, 775 and 776 didn't work, this means it needs read, write and execute permissions to "other users". 

Which makes me wonder how can I know which user needs this. I know it is not the `node` user because I changed ownership to the node user but it gave the same error: 
`RUN mkdir -p /app/.next/server`
`RUN chown -R node /app/.next/server`
`RUN chmod -R 770 /app/.next/server`

and I was sure the the user is node:
`USER node`

**Reply by mo-alaa** [reply]
Of course the above solution opens a security concern! So it doesn't solve the problem.
I also wonder how does it give that error while the application actually revalidates and gets the new updates from CMS!

**Reply by chandra2607** [reply]
I was facing similar persistent permission denied issue when my Next.js application tried to write image cache files in a Docker container. The error looked like this:

Initially, I tried setting `chmod 777` permissions which worked(as per suggestions by @mo-alaa ), but I knew this wasn't secure. When I changed it back to `chmod 755`, the errors returned. After hours of debugging, I discovered the root cause was a **user mismatch** between my Docker Compose configuration and Dockerfile.

## The Root Cause

Here's what was happening in my setup:

**My docker-compose.yml had:**
```yaml
services:
  nextjs-prod:
    user: 1000:1000  # This sets the container to run as node user
```

**My Dockerfile had:**
```dockerfile
RUN adduser --system --uid 1001 nextjs
USER nextjs  # This tries to switch to nextjs user
```

**File ownership was set to:**
```dockerfile
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
RUN chown -R nextjs:nodejs /app
```

The problem was that Docker Compose was overriding the Dockerfile's `USER` directive, so the actual running process was the `node` user (UID 1000) trying to write to files owned by `nextjs` user (UID 1001).

## How I Diagnosed It

I ran this command inside my running container to check what user was actually running the process:

```bash
docker exec -it nextjs-prod-1 ps aux
```

The output showed:
```
PID   USER     TIME  COMMAND
  1   node     0:01 next-server (v14)
 19   node     0:00 /bin/sh
 25   node     0:00 ps aux
```

This confirmed that despite my Dockerfile setting `USER nextjs`, the actual process was running as the `node` user due to the `user: 1000:1000` override in docker-compose.yml.

## The Solution

I had three options to fix this mismatch:

### Option 1: Remove the user override from docker-compose.yml
I commented out the user specification to let the Dockerfile control the user:

```yaml
services:
  nextjs-prod:
    # user: 1000:1000  # Removed this line
```

### Option 2: Align the Dockerfile with docker-compose (What I chose)
Since my docker-compose was setting `user: 1000:1000` (the built-in `node` user), I updated my Dockerfile to use the same user:

```dockerfile
# Removed custom user creation
# RUN adduser --system --uid 1001 nextjs

# Changed ownership to match the running user
COPY --from=builder --chown=node:node /app/.next/standalone ./
COPY --from=builder --chown=node:node /app/public ./public
COPY --from=builder --chown=node:node /app/.next/static ./.next/static
COPY --from=builder --chown=node:node /app/.next/cache ./.next/cache

RUN chown -R node:node /app
USER node
```

### Option 3: Update docker-compose to match Dockerfile
Alternatively, I could have updated docker-compose.yml to use the nextjs user:

```yaml
services:
  nextjs-prod:
    user: 1001:1001  # Match the nextjs user from Dockerfile
```

## Verification

After implementing the fix, I verified it worked by checking both the running process user and file ownership:

```bash
# Check what user is running the process
docker exec -it nextjs-prod-1 ps aux

# Check file ownership
docker exec -it nextjs-prod-1 ls -la /app/.next/
```

Both commands now showed the same user (`node`), and the permission errors disappeared. I was able to use secure `755` permissions instead of the overly permissive `777`.

**Comment by 8lane** [comment]
Seeing this exact issue as well with NextJs `13.4.12` using a mix of app router and pages router. All static routes give the `Failed to update prerender cache` error on first load after running the docker container.

**Comment by biohackerellie** [comment]
adding on to this, in my project also deployed in docker fully using app router, I'm getting similar errors 
```
 errno: -13,
  code: 'EACCES',
  syscall: 'mkdir',
  path: '/app/.next/cache/fetch-cache
```
and i do have this in my dockerfile
```
RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/public ./public
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static
COPY --from=builder --chown=nextjs:nodejs /app/.next/cache ./.next/cache
USER nextjs
```

**Comment by ghost** [comment]
Did you happen to find any futher updates on this? Experiencing currently with `13.4.19` when running containerized build.

**Reply by mo-alaa** [reply]
No updates yet, unfortunately.

**Comment by teguh77** [comment]
this error still exists, Does anyone already know the solution?

**Comment by taymoork2** [comment]
So a potential solution could be that your docker/kubernetes setup is not configured to run as a NonRoot user along with the correct nonroot user id for it to use.
This seemed to be the issue with my orgs internal infra setup that some configurations weren't setup for my specific deployments to use a nonrootuser

**Comment by sachin36987** [comment]
Adding the following lines before copying from builder and chown fixed the issue for me

RUN mkdir .next
RUN chown nextjs:nodejs .next

**Comment by Benman2005** [comment]
Ok, this is silly, but none of this worked for me, since i was dumb enough to not check my k8s settings file where it said `runAsUser: 1000` After changing it to the nextjs user id 1001 this error did not happen anymore :)

**Comment by DonikaV** [comment]
I have this now in the latest version 15.3.4 :|
